train_skypilot.py

```diff
@@ -0,0 +1,327 @@
+
+import sky
+
+from functionary.skypilot_utils import (
+    CLOUD_MAPPING,
+    check_features,
+    form_setup,
+    get_cloud_provider,
+)
+
+# Set up logging
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+logger = logging.getLogger(__name__)
+
+
+def parse_train_command(train_command_file: str) -> dict:
+    """
+    Parse a training command file and extract arguments into a dictionary.
+
+    This function reads a shell script containing training commands and parses the command line
+    arguments into a dictionary format. It handles both flag arguments (--flag) and key-value
+    arguments (--key value).
+
+    Args:
+        train_command_file (str): Path to the shell script containing the training command.
+
+    Returns:
+        dict: Dictionary mapping argument names (without -- prefix) to their values.
+              For flag arguments, the value will be True.
+              For key-value arguments, the value will be the string value provided.
+
+    Example:
+        For a command file containing:
+        ```
+        deepspeed train.py --model_name model1 --bf16 --epochs 3
+        ```
+        Returns:
+        {
+            'model_name': 'model1',
+            'bf16': True,
+            'epochs': '3'
+        }
+    """
+    # Read train command file and parse arguments into dictionary
+    train_args = {}
+    with open(train_command_file, "r") as f:
+        cmd = f.read()
+        # Split on newlines and backslashes to get all parts
+        parts = [p.strip() for p in cmd.replace("\\\n", " ").split()]
+        for i, part in enumerate(parts):
+            if part.startswith("--"):
+                key = part[2:]  # Remove -- prefix
+                # If this is the last argument or next is another flag, treat as boolean flag
+                if i == len(parts) - 1 or parts[i + 1].startswith("--"):
+                    train_args[key] = True
+                else:
+                    # Otherwise take the next part as the value
+                    train_args[key] = parts[i + 1]
+    return train_args
+
+
+def form_command(train_file_args: Dict[str, Any]) -> str:
+    """
+    Form the command string for training and model uploading.
+
+    This function constructs a multi-step command string that:
+    1. Changes to the functionary directory
+    2. Creates a training script with the provided arguments
+    3. Runs the training script
+    4. If using LoRA, merges the learned weights with the base model
+    5. Creates a private repo on Hugging Face
+    6. Uploads the trained model to Hugging Face
+
+    Args:
+        train_file_args (Dict[str, Any]): Dictionary of training arguments parsed from
+            the training command file, containing keys like 'model_name_or_path',
+            'output_dir', etc.
+
+    Returns:
+        str: The complete command string to execute all training and upload steps.
+    """
+    command = "cd functionary && "
```

Comment:
Skypilot does not open runpod instances with volume disks. It will allocate `args.disk_size` memory to the container disk. Provide appropriate `args.disk_size` values for the model that you want to use instead.